Add custom deployment role documentation (#3437)
astronomer/docs: astro/customize-deployment-roles.md

---

> title: "Create and assign custom roles for Astro Deployments"
> id: customize-deployment-roles

Reviewer: ```suggestion
title: "Create and assign custom roles for Astro Deployments"
sidebar-label: "Create and assign custom roles"
id: customize-deployment-roles
```
The linebreak in the sidebar is a bit noisy

description: Customize your users' permissions for Airflow environments on Astro.
---

:::privatepreview
:::

> :::caution

Reviewer: ```suggestion
:::info
```
We've got this as info level in the Reporting dashboard note - do we want them to both be caution?

Author: No I think info is probably better

This feature is available only in the Enterprise tier.

:::

A user entity's Deployment role determines their level of access to a specific Deployment in a Workspace. User entities with Workspace Member permissions or higher have some level of access to all Deployments in a Workspace, and these permissions can be heightened using Deployment roles.

> There are some circumstances where users should have limited access to Deployments in a Workspace. For example, all users might need full access to a development Deployment, but only administrative users should have access to a production Deployment.

Reviewer: ```suggestion
There are some circumstances where users should have limited access to Deployments in a Workspace. For example, all users might need full access to a development Deployment, but only administrative users need access to a production Deployment.
```

> In situations where you need fine-grained Deployment access, you can create custom Deployment roles and assign them to users with Workspace Accessor or Workspace Member roles. When you grant a user a Deployment role, they have a specific level of access to a specific Deployment. Use custom Deployment roles to enable users to collaborate in the same Workspace only with the minimum permissions they require.

Reviewer: ```suggestion
In situations where you need fine-grained Deployment access, you can create custom Deployment roles and assign them to users with Workspace Accessor or Workspace Member roles. When you grant a user a Deployment role, they have a specific level of access to a specific Deployment. Use custom Deployment roles to enable users to collaborate in the same Workspace, with only the minimum permissions they require.
```

## Prerequisites

- Organization Owner permissions to create, update, and delete custom roles.
- Workspace Owner permissions or Deployment Admin permissions to assign and change Deployment roles for users.

## Create a custom Deployment role

> Custom Deployment roles are created and managed at the Organization level. After you create a custom Deployment role, you can assign users, teams, and Deployment API tokens the role from any Deployment in the Organization. 

Reviewer: ```suggestion
Manage and create Custom Deployment roles at the Organization level. After you create a custom Deployment role, you can assign users, teams, and Deployment API tokens the role from any Deployment in the Organization. 
```

> 1. In the Cloud UI, click your Workspace name in the upper left corner, then click **Organization Settings**. 

Reviewer: ```suggestion
1. In the Astro UI, click your Workspace name in the upper left corner, then click **Organization Settings**. 
```
Have we made the switch yet?

   
2. Go to **Access Management**, then click **Roles**.

3. Click **Custom** then click **+ Add Role**.

4. In the window that appears, give your new role a **Name** and **Description**.

5. In the **Permissions** table, check the boxes of all permissions that you want the new role to have. See [Deployment role permission reference](deployment-role-reference.md) for more information about each available permission. 

    :::tip

    Use the dropdown menu above the permissions table to automatically load the permissions of a templated role or an existing custom role as the basis for your new role. See [Deployment role templates](#deployment-role-templates) for more information about the available default templates.

    :::

6. Click **Create role**.

> Your role is now available to assign at the Deployment level. See [Assign users and Teams to Deployments](#assign-users-and-teams-to-deployments) / [Create Deployment API tokens with custom Deployment roles](#create-deployment-api-tokens-with-custom-deployment-roles).

Reviewer: ```suggestion
Your role is now available to assign at the Deployment level. See [Assign users and Teams to Deployments](#assign-users-and-teams-to-deployments) or [Create Deployment API tokens with custom Deployment roles](#create-deployment-api-tokens-with-custom-deployment-roles) for next steps.
```

---

> title: "Create and assign custom roles for Astro Deployments"

Author: ```suggestion
title: "Create and assign custom Deployment roles"
```

    :::

6. Click **Create role**.

Your role is now available to assign at the Deployment level. See [Assign users and Teams to Deployments](#assign-users-and-teams-to-deployments) / [Create Deployment API tokens with custom Deployment roles](#create-deployment-api-tokens-with-custom-deployment-roles).

### Deployment role templates

Astro provides a few Deployment role templates that you can use as the basis for custom roles. These roles are not hard-coded and exist only as templates. 

- **Deployment Viewer**: This is similar to the [Airflow Viewer](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#viewer) viewer role. It grants the user entity view-only permissions for the Airflow UI excluding the **Admin** tab.
- **Deployment Author**: This is similar to the [Airflow User](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#user) role. It grants the user entity permissions to deploy code and manage DAG and task runs from the Airflow UI.

> - **Deployment Operator**: This is similar to the [Airflow Op](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#op) role. It grants the user entity permissions to update Deployment API tokens and Airflow objects from both the Airflow UI and the Cloud UI.

Author: ```suggestion
- **Deployment Operator**: This is similar to the [Airflow Op](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#op) role. It grants the user entity permissions to update Deployment API tokens and Airflow objects from both the Airflow UI and the Astro UI.
```

## Assign users and Teams to Deployments

Using Deployment roles, you can add users and Teams directly to Deployments without first assigning them to a Workspace. If they don't already belong to the Workspace, Astro grants them the _Workspace Accessor_ role. A Workspace Accessor only has permissions to access their assigned Deployments within the Workspace. All other Deployments and Workspace settings are hidden.

> 1. In the Cloud UI, open the Deployment where you want to assign the user entity.

Author: ```suggestion
1. In the Astro UI, open the Deployment where you want to assign the user entity.
```

---
title: "Create and assign custom roles for Astro Deployments"
id: customize-deployment-roles
unlisted: true
description: Customize your users' permissions for Airflow environments on Astro.
---

:::caution

> This feature is in [Private Preview](feature-previews.md).

Author: Core documentation to review

:::

> A user's Deployment role determines their level of access to a specific Deployment in a Workspace. By default, users, Teams, and Deployment API tokens that have Workspace Author permissions or above have Deployment Admin permissions to all Deployments in the Workspace.

Reviewer 2: `By default, users, Teams, and Deployment API tokens that have Workspace Author permissions or above have Deployment Admin permissions to all Deployments in the Workspace.`

This is not true. Deployment creators are granted Deployment Admin role, and the default role for Deployment API Tokens is Deployment Admin.

There are some circumstances where users should have limited access to Deployments in a Workspace. For example, all users might need full access to a development Deployment, but only administrative users should have access to a production Deployment.

> In situations where you need fine-grained Deployment access, you can create custom Deployment roles and assign them to users with Workspace Accessor or Workspace Viewer permissions. When you grant a user a Deployment role, they have a specific level of access to a specific Deployment. Use custom Deployment roles to enable users to collaborate in the same Workspace only with the minimum permissions they require.

Reviewer 2: ```suggestion
In situations where you need fine-grained Deployment access, you can create custom Deployment roles and assign them to users with Workspace Accessor or Workspace Member roles. When you grant a user a Deployment role, they have a specific level of access to a specific Deployment. Use custom Deployment roles to enable users to collaborate in the same Workspace only with the minimum permissions they require.
```

## Prerequisites

> - Organization Owner permissions

Reviewer 2: Org Owner to create, update, and delete custom roles. Can also assign them of course.
Workspace Owner, Deployment Admin, or users with a custom deployment role that has deployment access management permissions, can assign, change, or remove deployment roles from users, teams and tokens.

## Create a custom Deployment role

> Custom Deployment roles are created and managed at the Organization level. After you create a custom Deployment role, you can assign users the role from any Deployment in the Organization. 

Reviewer 2: ```suggestion
Custom Deployment roles are created and managed at the Organization level. After you create a custom Deployment role, you can assign users, teams, and Deployment API tokens the role from any Deployment in the Organization. 
```

1. In the Cloud UI, click your Workspace name in the upper left corner, then click **Organization Settings**. 
   
2. Go to **Access Management**, then click **Roles**.

3. Click **Custom** then click **+ Add Role**.

4. In the window that appears, give your new role a **Name** and **Description**.

5. In the **Permissions** table, check the boxes of all permissions that you want the new role to have. See [Deployment role permission reference](deployment-role-reference.md) for more information about each available permission. 

    :::tip

    Use the dropdown menu above the permissions table to automatically load the permissions of a templated role or an existing custom role as the basis for your new role. See [Deployment role templates](#deployment-role-templates) for more information about the available default templates.

    :::

6. Click **Create role**.

> Your role is now available to assign at the Deployment level. See [Assign users, Teams, and API tokens to Deployments](#assign-users-teams-and-api-tokens-to-deployments).

Reviewer 2: I think this link is broken

> Your role is now available to assign at the Deployment level. See [Assign users, Teams, and API tokens to Deployments](#assign-users-teams-and-api-tokens-to-deployments).

Author: ```suggestion
Your role is now available to assign at the Deployment level. See [Assign users and Teams to Deployments](#assign-users-and-teams-to-deployments) / [Create Deployment API tokens with custom Deployment roles](#create-deployment-api-tokens-with-custom-deployment-roles).
```

### Deployment role templates

Astro provides a few Deployment role templates that you can use as the basis for custom roles. These roles are not hard-coded and exist only as templates. 

> - **Deployment Viewer**: This is similar to the [Airflow viewer](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#viewer) viewer role. It grants the user entity view-only permissions for the Airflow UI excluding the **Admin** tab.

Reviewer 2: ```suggestion
- **Deployment Viewer**: This is similar to the [Airflow Viewer](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#viewer) viewer role. It grants the user entity view-only permissions for the Airflow UI excluding the **Admin** tab.
```

> - **Deployment Author**: This is similar to the [Airflow user](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#user) role. It grants the user entity permissions to deploy code and manage DAG and task runs from the Airflow UI.

Reviewer 2: ```suggestion
- **Deployment Author**: This is similar to the [Airflow User](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#user) role. It grants the user entity permissions to deploy code and manage DAG and task runs from the Airflow UI.
```

- **Deployment Operator**: This is similar to the [Airflow Op](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#op) role. It grants the user entity permissions to update Deployment API tokens and Airflow objects from both the Airflow UI and the Cloud UI.

## Assign users and Teams to Deployments

> Before you assign a custom Deployment role to a user or Team, ensure that the user entity doesn't have Workspace Admin or Workspace Member permissions. Otherwise, the user entity will already have all available Deployment permissions and the Deployment role will have no effect.

Reviewer 2: I think this statement is only true for Workspace Admin. While other Workspace roles have broad permissions to all Deployments, a custom Deployment role can still be considered additive depending on the Workspace role and custom Deployment role combination.

Reviewer 2: For example, granting Workspace Member to users gives them read only access to all Deployments, and then a privileged custom Deployment role to a subset of users for a subset of Deployments.

Author: Got it, just going to delete this line then. I thought there was a greater risk of over-granting permissions to users, but it seems like almost every Workspace role can have heightened permissions with Deployment roles

1. In the Cloud UI, open the Deployment where you want to assign the user entity.
2. Click **Access**, then click **Users** or **Teams** depending on what kind of user entity you want to assign to the Deployment.

> 3. Click **+ User**/ **+ Team**.

Reviewer 2: ```suggestion
3. Click **+ User** / **+ Team**.
```

Author: This looks awkward in the raw markdown, but it's our style to only add a space after a forward slash

4. In the window that appears, select the user entity you want to add, then select the role they will have in the Deployment.

> 5. Click **Add User**/ **Add Team**.

Reviewer 2: ```suggestion
5. Click **Add User** / **Add Team**.
```

## Create Deployment API tokens with custom Deployment roles

By default, Deployment API tokens have Deployment Admin permissions. You can alternatively create or update Deployment API tokens to use any of your configured custom Deployment roles. 

1. In the Cloud UI, open the Deployment where you want to assign the user entity.
2. Click **Access**, then click **API Tokens**.
3. Click **+ Deployment API Token**.
4. In the window that appears, select give your new Deployment API key a name, custom role, and expiration date.

> 5. Click **Add User**/ **Add Team**.

Reviewer 2: I don't believe this is correct

### Deployment role templates

Astro provides a few Deployment role templates that you can use as the basis for custom roles. These roles are not hard-coded and exist only as templates. 

- **Deployment Viewer**: This is similar to the [Airflow Viewer](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#viewer) viewer role. It grants the user entity view-only permissions for the Airflow UI excluding the **Admin** tab.
- **Deployment Author**: This is similar to the [Airflow User](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#user) role. It grants the user entity permissions to deploy code and manage DAG and task runs from the Airflow UI.
- **Deployment Operator**: This is similar to the [Airflow Op](https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#op) role. It grants the user entity permissions to update Deployment API tokens and Airflow objects from both the Airflow UI and the Cloud UI.

## Assign users and Teams to Deployments

> Before you assign a custom Deployment role to a user or Team, ensure that the user entity doesn't have Workspace Admin or Workspace Member permissions. Otherwise, the user entity will already have all available Deployment permissions and the Deployment role will have no effect.

Author: ```suggestion
```

1. In the Cloud UI, open the Deployment where you want to assign the user entity.
2. Click **Access**, then click **Users** or **Teams** depending on what kind of user entity you want to assign to the Deployment.
3. Click **+ User**/ **+ Team**.
4. In the window that appears, select the user entity you want to add, then select the role they will have in the Deployment.
5. Click **Add User**/ **Add Team**.

> ## Create Deployment API tokens with custom Deployment roles

Author: FYI this topic will be deleted when the feature is public, as it will be covered in the Deployment API token docs.

By default, Deployment API tokens have Deployment Admin permissions. You can alternatively create or update Deployment API tokens to use any of your configured custom Deployment roles. 

1. In the Cloud UI, open the Deployment where you want to assign the user entity.
2. Click **Access**, then click **API Tokens**.
3. Click **+ Deployment API Token**.
4. In the window that appears, select give your new Deployment API key a name, custom role, and expiration date.

> 5. Click **Add User**/ **Add Team**.

Author: ```suggestion
5. Click **Add Deployment API token**.
```
